### PR TITLE
[RW-384] Profile storage service

### PR DESCRIPTION
**Pull request by blrubenstein**

### CODE REVIEW

File: ui/src/app/services/profile-storage.service.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+@Injectable()
+export class ProfileStorageService {
+  private activeCall = false;
+  private profile = new ReplaySubject<Profile>(1);
+  public profile$ = this.profile.asObservable();
+
+  constructor(private profileService: ProfileService) {}
+
+  requestNewProfile() {
```
Comment (calbach) [inline ui/src/app/services/profile-storage.service.ts]: Perhaps name this `reload()`, then (maybe) dispatch the initial `getMe()` on service creation (possibly blocked on signin)? This would allow us to satisfy components that would just take a stale view of the profile, as well as those that want to force a reload. e.g. the bug report button probably shouldn't need to force a Profile reload.

File: ui/src/app/services/profile-storage.service.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+import {Profile} from 'generated';
+import {ProfileService} from 'generated';
+
+
+@Injectable()
+export class ProfileStorageService {
+  private activeCall = false;
+  private profile = new ReplaySubject<Profile>(1);
+  public profile$ = this.profile.asObservable();
```
Comment (calbach) [inline ui/src/app/services/profile-storage.service.ts]: Unless this is following some existing convention, would prefer not to use `$` in the property name. Could be `profile` or `profileObs`?

File: ui/src/app/views/profile-page/component.ts [hunk 3]
```diff
@@ -46,19 +47,24 @@ export class ProfilePageComponent implements OnInit {
...
   ngOnInit(): void {
     this.profileImage = this.signInService.profileImage;
     this.errorText = null;
-    this.profileService.getMe().subscribe(
-      (profile: Profile) => {
-        this.profile = profile;
+
+    this.profileStorageService.profile$.subscribe(profile => {
+      this.profile = profile;
+      if (this.editing === false) {
```
Comment (calbach) [inline ui/src/app/views/profile-page/component.ts]: `!this.editing`

File: ui/src/app/views/workspace-list/component.ts [hunk 4]
```diff
@@ -68,15 +71,7 @@ export class WorkspaceListComponent implements OnInit, OnDestroy {
...
-    this.profileService.getMe().subscribe((profile: Profile) => {
-      if (profile.freeTierBillingProjectStatus === BillingProjectStatus.Ready) {
-        this.billingProjectInitialized = true;
-      } else {
-        this.billingProjectQuery = setTimeout(() => {
-          this.queryBillingStatus();
-        }, 10000);
-      }
-    });
+    this.profileStorageService.requestNewProfile();
```
Comment (calbach) [inline ui/src/app/views/workspace-list/component.ts]: The wrapper method doesn't seem necessary any longer, would just inline this above.

File: ui/src/app/views/signed-in/component.ts [hunk 5]
```diff
@@ -51,18 +52,21 @@ export class SignedInComponent implements OnInit {
   ngOnInit(): void {
 
     this.profileImage = this.signInService.profileImage;
```
Comment (calbach) [inline ui/src/app/views/signed-in/component.ts]: Not really part of this PR, but why would this value be available before the user has signed in?

File: ui/src/app/services/profile-storage.service.ts (hunk 2 above)
Comment (rushtong) [reply ui/src/app/services/profile-storage.service.ts]: I've read that the `$` is convention to denote an observable in angular (https://cycle.js.org/basic-examples.html#basic-examples-increment-a-counter-what-is-the-convention). I'm still ramping up on the angular world, so this is an odd style for me as well.

File: ui/src/app/views/signed-in/component.ts [hunk 6]
```diff
@@ -51,18 +52,21 @@ export class SignedInComponent implements OnInit {
...
+
+    this.profileStorageService.profile$.subscribe((profile) => {
+      this.hasReviewResearchPurpose =
+        profile.authorities.includes(Authority.REVIEWRESEARCHPURPOSE);
+      this.hasReviewIdVerification =
+        profile.authorities.includes(Authority.REVIEWIDVERIFICATION);
+      // this.email = profile.username;
```
Comment (rushtong) [inline ui/src/app/views/signed-in/component.ts]: nit: remove if not needed.

File: ui/src/app/services/profile-storage.service.ts (hunk 2 above)
Comment (blrubenstein) [reply ui/src/app/services/profile-storage.service.ts]: Yeah, when I was working on the `SignInService` John told me about that convention. I would be happy to follow a different convention if we wanted to use another one.

File: ui/src/app/services/profile-storage.service.ts (hunk 1 above)
Comment (blrubenstein) [reply ui/src/app/services/profile-storage.service.ts]: Sounds good! Will do.

File: ui/src/app/views/signed-in/component.ts (hunk 5 above)
Comment (blrubenstein) [reply ui/src/app/views/signed-in/component.ts]: Fair, will move.

**Comment by blrubenstein** [comment]
I could picture the profile-storage-service turning into a generic caching service at some point also.

File: ui/src/app/services/profile-storage.service.ts (hunk 2 above)
Comment (johnstill) [reply ui/src/app/services/profile-storage.service.ts]: I don't generally like Hungarian notation but we pretty frequently have a source and an observable view on that source (that is the case here in fact), and the dollar suffix is a visible and succinct way to indicate which is which.

**Comment by calbach** [comment]
+1, I think we could do away with the bug report, workspace page, and workspace share page reloads in particular.

Also don't wait on my approval, I'm out today and tomorrow.
